### ISSUE TITLE
Riemann data: build theorem atlas, relevance graph, corpora, and holdouts

### ISSUE THREAD

**Issue by murillo128** [murillo128/qwen-lean#37]
Parent epic: #36

## Readiness
**Current state:** in-progress

This is the **single controlling execution issue for the complete pre-training data/discovery stage** of the Riemann deep-specialization branch.

One Codex executor should take this issue end to end and deliver one ready-for-review PR. Do not create child issues merely to split the stages below; create a separate blocker only for genuinely independent work that cannot be completed in this execution.

No GPU training is required. Do not wait for #27 to finish.

## Goal

Produce one repository-owned, reproducible Riemann/prime-number-theory data layer that contains:

1. the accepted Phase 2 mathlib corpus as a canonical versioned repository dataset;
2. an auditable formal relevance graph over that corpus;
3. a broader theorem atlas covering external Lean, other proof assistants, and mathematical literature;
4. verified specialist training corpora plus clean validation/near/far holdouts;
5. a ranked map of important missing formalizations and external Lean material worth ingesting later.

The key distinction is **mathematical relevance vs Lean-training readiness**. Literature statements, another prover's proof, or unverified Lean translations must never enter Lean SFT corpora as if they were verified Lean examples.

---

# Resolved implementation solution

## A. Remove the hidden Phase 2 artifact dependency

The accepted Phase 2 full corpus is not currently committed to Git: `main` contains compact `evidence/phase2/*`, while the root README documents that `train.jsonl`, `validation.jsonl`, and `heldout.jsonl` live under ignored `artifacts/` and can be regenerated with the committed extractor.

A fresh Codex checkout therefore must **not** assume that `artifacts/phase2/mathlib-whole-proof-v1` already exists.

Use this recovery contract:

1. If a local Phase 2 artifact exists, validate it against `config/phase2-mathlib.json` and `evidence/phase2/manifest.json` before using it.
2. If it does not exist, regenerate it using the exact committed Phase 2 command/toolchain from the root README and the pinned miniF2F checkout.
3. The regenerated corpus is accepted only if it reproduces the material Phase 2 identity and invariants below.
4. Once validated, **promote the accepted Phase 2 corpus itself into a canonical repository-owned dataset in this PR**. This is a packaging/versioning change, not a new Phase 2 model-selection experiment.
5. All Riemann internal corpora must then derive from this repository-owned snapshot, not from an executor-private artifact path.

### Frozen Phase 2 identity

Use the existing committed configuration/evidence as source of truth:

- dataset schema: `mathlib-whole-proof-v1`;
- mathlib revision: `81a5d257c8e410db227a6665ed08f64fea08e997` (`v4.32.0`);
- Lean: `leanprover/lean4:v4.32.0`;
- LeanDojo-v2 revision: `baed5eae6e87a65a446d9f54af07aab2154e7599`;
- miniF2F contamination reference: `f0a20e14c1eeccd859d51bb4c2b3ee487889c303`;
- tokenizer revision: `Qwen/Qwen3-8B-Base@49e3418fbbbca6ecbdf9608b4d22e5a407081db4`;
- expected final records: **88,958**;
- train / validation / heldout: **80,062 / 4,448 / 4,448**;
- remaining exact miniF2F statement matches: **0**;
- cross-split files/components/statement fingerprints: **0**.

A material mismatch is an integrity failure. Do not silently create a new Phase 2 corpus under the old name.

## B. Repository data layout

**Every authoritative versioned corpus produced or consumed as a project corpus by this stage must land on `main`.** A fresh checkout after merge must not require the executor's old `artifacts/` directory.

Use a simple repository-owned layout, for example:

```text
data/
  mathlib-whole-proof-v1/
    manifest.json
    train/...
    validation/...
    heldout/...
  riemann/
    sources.json
    seeds.json
    atlas.jsonl
    relations.jsonl
    internal-graph/...
    number-theory-wide-v1/...
    riemann-bubble-v1/...
    riemann-core-v1/...
    specialist-validation-v1/...
    riemann-near-holdout-v1/...
    number-theory-far-holdout-v1/...
    riemann-external-lean-v1/...   # only if accepted
```

Exact paths may differ if equally clear.

Prefer JSON/JSONL. If size makes plain files impractical, use deterministic sharding and deterministic compression (for example gzip with fixed metadata) with per-shard hashes in the manifest. Keep individual files comfortably below GitHub's hard file-size boundary. Do not use an ignored local artifact as the authoritative copy.

Third-party checkouts, tracing caches, tokenizer/model caches, full paper caches, temporary databases, and bulky logs remain outside Git.

## C. One source-normalization pipeline

Implement one small reusable data model rather than unrelated scripts for each source.

Every theorem/atlas record should expose, when applicable:

- stable project record id;
- source kind: `mathlib`, `external-lean`, `other-prover`, `literature-only`;
- repository/publication identity and immutable revision/citation;
- declaration/theorem canonical name and aliases;
- statement;
- verified Lean proof/completion only when genuinely available;
- file/source identity;
- license/provenance;
- mathematical topic/family tags;
- formal relation to RH;
- formalization/verification status;
- candidate role: `training-source`, `evaluation-target`, `formalize-in-lean`, `knowledge-only`.

Do not force literature or another prover into the Lean-training schema. The unified atlas can reference them; verified Lean corpus records remain a stricter subset.

## D. Internal relevance graph solution

Build the internal graph over the accepted Phase 2 records using existing resolved-premise metadata plus source identity.

Typed internal relations:

- `premise`: theorem A directly uses B;
- `user`: reverse index of `premise`;
- `same-source`: source/module neighborhood, weaker than a proof dependency;
- Phase 2 fingerprint/component identity for leakage control.

Discover and commit a reviewed seed manifest covering at least:

- primes / prime arithmetic;
- prime-counting and PNT-related declarations;
- Riemann zeta;
- Dirichlet/L-series and nonvanishing machinery;
- Möbius / von Mangoldt / arithmetic functions;
- directly connected real/complex analytic machinery.

Materialize deterministic relevance classes:

- `core`;
- `premise-1`;
- `premise-2`;
- `user-1`;
- `user-2`;
- `source-neighborhood`;
- `number-theory-control`.

Default graph depth is two dependency hops in either direction. Generic high-degree infrastructure nodes may be retained as prerequisites but must not recursively explode the bubble. Implement a deterministic/configured hub-expansion guard, record the actual degree distribution/cutoff used, and freeze it before corpus membership is materialized. No model output or benchmark result may influence these rules.

Every selected record must retain provenance explaining **why** it entered the bubble: seed, relation type, shortest distance, source, original Phase 2 record id/split/fingerprint/component, and canonical class.

## E. External Lean solution

The first mandatory external Lean source is PNT+:

- repository: `AlexKontorovich/PrimeNumberTheoremAnd`;
- current inspected `main` head to pin for this execution: `7715064f690d0689f30889846f4e2c5e7ec0c47e`;
- license: Apache-2.0.

Record that SHA in the source manifest and use it throughout the accepted run unless an explicit pre-execution correction is needed. Do not float on `main` after extraction begins.

Inspect the actual pinned source/blueprint for PNT, strong/error-term PNT, zeta bounds and functional machinery, zero counting/zero-free regions, Hadamard factorization, Perron/Mellin machinery, explicit prime estimates, and other directly connected analytic-number-theory results.

Also search current Lean project/community indexes for additional relevant Lean repositories. PNT+ is the mandatory first source, not an assertion that it is the only source.

An external Lean theorem may enter `riemann-external-lean-v1` only when all of the following are recorded and pass:

- immutable source revision;
- license allows repository redistribution of the derived record/proof material;
- exact declaration/source/proof identity is recoverable;
- it verifies in its pinned native Lean project/environment;
- it is not an exact statement duplicate of a protected internal/miniF2F evaluation record;
- its provenance remains distinct from the Phase 2-derived corpus.

Do not require external theorems to compile under qwen-lean's mathlib revision merely to establish that the upstream Lean proof is genuine. Compatibility for eventual mixed training can be addressed later.

## F. Other prover + literature atlas solution

Build a **source-bounded census**, not a claim of global completeness.

Inspect relevant Isabelle, Metamath, Coq, HOL, or comparable primary formalization sources and high-quality mathematical literature. Surveys may discover results, but implication/equivalence claims should be traced to an attributable mathematical source where practical.

Classify mathematical relationships to RH using typed relations such as:

- `definition/foundation`;
- `prerequisite`;
- `equivalent-to-RH`;
- `RH-implies`;
- `implies-RH`;
- `conditional-on-RH`;
- `partial-progress`;
- `analogue/generalization`;
- `heuristic/conjectural-neighborhood`.

Cover the major families around `π`, `θ`, `ψ`, von Mangoldt/PNT error formulations; Möbius/Mertens criteria; arithmetic-function/divisor-sum criteria; zeta/xi zeros and zero-free regions; explicit formulae; derivative/zero criteria; approximation/functional-analytic criteria; finite-height/conditional explicit estimates; Dirichlet L-functions and GRH analogues.

Every `equivalent-to-RH`, `RH-implies`, or `implies-RH` edge needs a source. Topical similarity alone is not a logical edge.

Commit only the project atlas/metadata/concise normalized statements and citations required by the project; do not vendor full copyrighted papers/books into the repo.

## G. Specialist corpus membership solution

Use one set of frozen relevance rules across all original Phase 2 splits.

### Training corpora — Phase 2 `train` only

Materialize complete records, not ID-only manifests:

- `number-theory-wide-v1`: all Phase 2 **train** records satisfying the broad number-theory selector;
- `riemann-bubble-v1`: all Phase 2 **train** records satisfying the frozen Riemann bubble rules;
- `riemann-core-v1`: the tight highest-relevance Phase 2 **train** subset.

Do not duplicate rows for later oversampling; preserve class/relevance metadata so training can assign weights later.

External Lean records, if accepted, stay in the separate `riemann-external-lean-v1` provenance lane rather than being silently mixed into the internal pool.

### Specialist validation — Phase 2 `validation`

`specialist-validation-v1` is the set of Phase 2 **validation** records selected by the same number-theory/Riemann relevance rules. It may later be used for teacher-forced diagnostics/checkpoint selection but never optimizer batches.

### Near holdout — Phase 2 `heldout`

`riemann-near-holdout-v1` is the set of Phase 2 **heldout** records that would qualify for `core`/`riemann-bubble` under the same frozen relevance rules.

### Far/domain holdout — Phase 2 `heldout`

`number-theory-far-holdout-v1` is the set of Phase 2 **heldout** records that qualify for broad number theory but are outside `riemann-near-holdout-v1`.

This preserves the already accepted Phase 2 semantics cleanly: train stays train, validation stays validation, heldout stays heldout. Do not invent a second random theorem-level split unless a demonstrated insufficiency requires a later design change.

For all lanes require:

- no forbidden exact statement-fingerprint duplication across train/validation/holdout;
- existing miniF2F exact-overlap hygiene remains intact;
- external Lean training candidates are screened against protected evaluation fingerprints;
- dependency edges crossing train↔holdout are reported as proximity diagnostics, not silently removed unless they violate an explicit split rule;
- membership is frozen before specialist model results exist.

The RH statement itself and other unproved conjectures are knowledge/research anchors, not ordinary proof-success targets.

## H. Outputs committed by the PR

At minimum `main` must contain after merge:

- canonical `mathlib-whole-proof-v1` corpus + manifest;
- source manifest;
- reviewed seed manifest;
- internal relevance graph/provenance data;
- unified Riemann theorem atlas;
- typed mathematical relationship graph;
- `number-theory-wide-v1`;
- `riemann-bubble-v1`;
- `riemann-core-v1`;
- `specialist-validation-v1`;
- `riemann-near-holdout-v1`;
- `number-theory-far-holdout-v1`;
- `riemann-external-lean-v1` if its acceptance gates pass;
- coverage/statistics/leakage evidence;
- ranked lists of external-Lean ingestion, missing Lean formalization, future evaluation, and knowledge-only opportunities.

---

# Validation / hard gates

Before ready-for-review, prove at least:

1. Phase 2 input exists or was regenerated from the committed pinned pipeline.
2. Its identity matches the accepted 88,958 / 80,062 / 4,448 / 4,448 corpus and split/contamination invariants.
3. The canonical Phase 2 corpus can be loaded directly from a fresh repository checkout without the old local `artifacts/phase2` directory.
4. Internal graph construction and ring assignment are deterministic on hand-checkable fixtures.
5. Hub-expansion guards prevent generic dependencies from swallowing most of mathlib.
6. Every internal specialist row preserves authoritative Phase 2 statement/proof/source identity.
7. Specialist training corpora contain Phase 2 train records only.
8. Specialist validation comes from Phase 2 validation only; near/far holdouts come from Phase 2 heldout only.
9. No forbidden fingerprint/component leakage or miniF2F exact overlap is introduced.
10. PNT+ and any accepted external Lean source are pinned, licensed, provenance-complete, and natively Lean-verified.
11. External Lean, other-prover, literature, and mathlib statuses cannot be confused programmatically.
12. Mathematical relationship edges are typed separately from Lean premise edges.
13. Atlas aliases/dedup and corpus memberships are deterministic.
14. All authoritative corpus/atlas files expected by the manifests are present in Git and hash-valid.
15. Normal repository tests remain CPU-only; external source checkouts/caches are not required merely to load the committed final corpora.

Run repository-native tests/lint/format/diff checks and one final-capable independent review over the complete diff and data evidence.

A small or surprising Riemann corpus is a valid scientific result. Hidden provenance, leakage, unverified proofs, or a fresh checkout that cannot load the committed corpora are implementation failures.

# Out of scope

- choosing the final Qwen/foundation model;
- SFT/QLoRA training;
- GRPO/RL/verifier-reward training;
- automatically translating literature/other-prover proofs into Lean;
- trusting generated/unverified Lean proofs;
- formalizing the ranked missing theorems in this same PR;
- changing generic qwen-lean model lineage;
- claiming the atlas is globally exhaustive;
- claiming progress on proving RH from corpus/atlas coverage.

# Delivery

Deliver **one focused ready-for-review PR** with the bootstrap/repackaging of accepted Phase 2 data, internal graph/corpus pipeline, external-source discovery/indexing needed for accepted outputs, complete versioned corpora, atlas/relationship data, tests, configs/manifests, and compact analysis/evidence.

Do not merge automatically. After acceptance/merge, update #36 with actual corpus/ring/holdout sizes, external Lean coverage, major atlas gaps, and implications for specialist SFT design.

**Comment by murillo128** [comment]
## Amendment — corpus payloads are local repository data

This amendment is authoritative for #37 and supersedes any alternative interpretation based on GitHub Release assets, Git LFS pointers, remote object stores, or executor-private `artifacts/` directories.

### Decision

**Every accepted corpus payload must live physically in the normal local `qwen-lean` checkout and be versioned by ordinary Git on `main`.**

A fresh normal clone/checkout of the accepted commit must contain the actual corpus bytes required to inspect, load, evaluate, or train another model. Fetching a GitHub Release, Git LFS object, Hugging Face dataset, S3/object-store asset, or a previous executor's local artifact is not part of the normal corpus-loading contract.

This applies to the canonical Phase 2 snapshot promoted by #37 and to every authoritative corpus produced by this stage, including:

- `mathlib-whole-proof-v1`;
- `number-theory-wide-v1`;
- `riemann-bubble-v1`;
- `riemann-core-v1`;
- `specialist-validation-v1`;
- `riemann-near-holdout-v1`;
- `number-theory-far-holdout-v1`;
- `riemann-external-lean-v1` if accepted;
- theorem-atlas / relation / provenance datasets required to reproduce membership.

### Storage solution

Use normal repository files under the repository-owned `data/` layout. Prefer inspectable JSON/JSONL when practical.

If a logical corpus is too large for one convenient Git/GitHub object:

1. deterministically shard it into ordinary Git-tracked files;
2. optionally deterministically compress shards;
3. keep every individual tracked file safely below GitHub's per-file hard limit;
4. record ordered shard names, record counts, byte sizes, and SHA-256 hashes in the corpus manifest;
5. make the repository loader read the committed shards directly.

Do **not** replace corpus bytes with Git LFS pointer files. Do **not** require a Release download in the loader. A future Release may mirror an accepted dataset for convenience, but it is non-authoritative and the corpus must remain complete without it.

### Reproduction semantics

We now require both forms of reproducibility:

- **artifact reproduction:** checkout the accepted commit and obtain the exact accepted corpus bytes immediately;
- **derivation reproduction:** rerun the pinned Phase 2 / #37 source pipeline and compare the regenerated corpus against the committed manifests/hashes.

Regeneration is therefore a verification path, not the only way to recover training data.

During #37 execution, once a corpus is accepted/materialized, move/write its canonical form under the tracked `data/` tree and run subsequent materialization/tests against that repository-local form rather than continuing to depend on ignored `artifacts/` paths.

### Additional hard gates

Before ready-for-review, prove that:

- `git ls-files data/...` includes every manifest-declared corpus shard;
- a clean checkout with no `artifacts/`, no LFS fetch, and no network access can load all accepted corpus datasets;
- corpus manifests verify ordered record counts and SHA-256 hashes for all committed shards;
- rebuilding from pinned upstream sources either reproduces the accepted bytes/hashes or fails loudly as an integrity discrepancy;
- no training/evaluation entry point for these accepted datasets silently prefers an external/ignored copy over the committed repository-local corpus.

This does not change the rule that third-party source checkouts, LeanDojo traces/caches, tokenizer/model caches, papers, temporary databases, and bulky logs may remain outside Git. The distinction is: **source/cache/intermediate material may be external; accepted project corpus payloads may not.**
